statsmodels/stats/multicollinearity.py

```diff
@@ -0,0 +1,539 @@
+        self.mom = xm
+
+    @cache_readonly
+    def mom_inv(self):
+        """inverse of the moment matrix
+
+        cached, behavior in (near) singular cases comes from numpy.linalg.inv
+        """
+        return np.linalg.inv(self.mom)
+
+    def get_vif(self, ridge_factor=None):
+        """Variance inflation factor based on moment or correlation matrix.
+
+        Parameters
+        ----------
+        ridge_factor : float
+            A ridge factor is added to the moment matrix before inverting it.
+            This can be used to avoid problems with singular moment matrices.
+            By the default the ridge factor of the instance is used, but can
+            be overridden here.
+
+        Returns
+        -------
+        vif : ndarray
+            variance inflation factor
+        """
+        if ridge_factor is None:
+            ridge_factor = self.ridge_factor
+
+        if ridge_factor == 0:
+            mmi = self.mom_inv
+        else:
+            ridge = self.ridge_factor * np.eye(self.k_vars)
+            mmi = np.linalg.inv(self.mom + ridge)
+
+        # np.diag returns read only array, need copy
+        vif_ = np.diag(mmi).copy()
+        return vif_
+
+    @cache_readonly
+    def vif(self):
+        """Variance inflation factor based on moment or correlation matrix.
+
+        cached attribute, uses the ridge_factor in the attribute
+        """
+        vif_ = self.get_vif()
+        # It is possible that singular matrix has slightly negative eigenvalues
+        # and large negative instead of positive vif
+        mask = vif_ < 1e-13
+        vif_[mask] = np.inf
+        return vif_
+
+    @property
+    def tss(self):
+        """Total mean sum of squares in partial regression
+
+        Note: this is just ones for standardized data
+        """
+        return np.diag(self.mom)
+
+    @property
+    def rss(self):
+        """Residual mean sum of squares in partial regression
+
+        This is the residual mean squared error without degrees of freedom
+        correction for the partial regression when standardized data are used.
+        """
+        return 1 / np.diag(self.mom_inv)
+
+    @property
+    def rsquared_partial(self):
+        """Partial correlation of one variable with all others
+
+        This corresponds to R^2 in a linear regression of one variable on all
+        others, including a constant if standarize is true, i.e. data was
+        demeaned.
+
+        not cached
+        """
+        # direct computation, but vif is cached so we use that
+        # r2p = 1 - self.rss / self.tss
+        r2p = 1. - 1. / self.vif
+        return r2p
+
+    @property
+    def corr_partial(self):
+        """partial correlation matrix
+
+        Pairwise partial correlation of two variables conditional on remaining
+        variables.
+
+        Warning: This uses the matrix inverse of the moment matrix, which can
+        produce nans or raise exceptions in singular or near-singular cases.
```

Comment:
check is this exog-scale invariant, it should be

Reply:
possible problem here: I think this method assumes that mom_inv is demeaned to have an interpretation as correlation

(not checked, I'm working on semi-partial correlation and in an example I get a diff of around 1e-4 for semi-partial corr between corr and mom versions.